app/src/main/java/com/nutomic/syncthingandroid/service/EventProcessor.java

```diff
@@ -242,10 +209,80 @@ public void stop() {
         }
     }
 
-    private void notify(String text, PendingIntent pi) {
-        // HACK: Use a random, deterministic ID between 1000 and 2000 to avoid duplicate
-        //       notifications.
-        int notificationId = 1000 + text.hashCode() % 1000;
-        mNotificationHandler.showEventNotification(text, pi, notificationId);
+    private void onDeviceRejected(String deviceId, String deviceName) {
+        if (deviceId == null) {
+            return;
+        }
+        Log.d(TAG, "Unknown device " + deviceName + "(" + deviceId + ") wants to connect");
+
+        String title = mContext.getString(R.string.device_rejected,
+                deviceName.isEmpty() ? deviceId.substring(0, 7) : deviceName);
+        int notificationId = mNotificationHandler.getNotificationIdFromText(title);
+
+        // Prepare "accept" action.
+        Intent intentAccept = new Intent(mContext, DeviceActivity.class)
+                .putExtra(DeviceActivity.EXTRA_NOTIFICATION_ID, notificationId)
+                .putExtra(DeviceActivity.EXTRA_IS_CREATE, true)
+                .putExtra(DeviceActivity.EXTRA_DEVICE_ID, deviceId)
+                .putExtra(DeviceActivity.EXTRA_DEVICE_NAME, deviceName);
+        // Use a deterministic ID to make multiple PendingIntents distinguishable.
+        int requestCode = deviceId.hashCode();
+        PendingIntent piAccept = PendingIntent.getActivity(mContext, requestCode,
+            intentAccept, PendingIntent.FLAG_UPDATE_CURRENT);
+
+        // Prepare "ignore" action.
+        Intent intentIgnore = new Intent(mContext, SyncthingService.class)
+                .putExtra(SyncthingService.EXTRA_NOTIFICATION_ID, notificationId)
+                .putExtra(SyncthingService.EXTRA_DEVICE_ID, deviceId);
+        intentIgnore.setAction(SyncthingService.ACTION_IGNORE_DEVICE);
+        PendingIntent piIgnore = PendingIntent.getService(mContext, 0,
+            intentIgnore, PendingIntent.FLAG_UPDATE_CURRENT);
```

Comment:
How does this work, why do we have 3 intents, two pending, one non-pending etc? What's the difference here? I know nothing about android.

Reply:
It's okay like this because the intents are nothing that already "lives" at that time. You can see them as "prepared" for later use. That's why the two intents (accept, ignore) are put into a pending intent for each (accept, ignore) and then passed on to our NotificationHandler class (passing it on to android's notification service). Android docs say, that a PendingIntent in a notification is a set of "what should be executed" (e.g. activity or service action like it is here) and which the target object needs to know (that's why we put extras in every intent). Android also has a built-in mechanism to take care, intents that represent the same (told by the requestCode) are only kept one-time in memory and are internally referenced. As you asked in your other question, the requestCode should stay like this (compare to the "old code" that used the same approach for producing the same requestCode when the folder/device combination we talk about is the same per triggered event). I only added the PendingIntent.FLAG_UPDATE_CURRENT as the extra's are not carried along to the service/activity launched when hitting the action button instead of clicking in the middle of the notification field itself. That's specific to the action button interface of android's notification service since API level 16+ and does not harm our compatibility level.